File: FizzBuzz/kotlin/src/main/kotlin/FizzBuzz.kt
object FizzBuzz {

	private var printers: MutableList<(List<String>) -> String> = mutableListOf()

	fun stringFor(input: List<Any>): String
	{
		val result = input.map({ when(it)
		{
			is Int -> ProcessJob.process(it)
			else -> ""
		}})
		return print(result)
	}

	fun stringFor(input: Int): String
	{
		val result = ProcessJob.process(input)
		return print(listOf(result))
	}

	fun addPrinter(printer: (List<String>) -> String)
	{
		printers.add(printer)
	}

	fun removeAllPrinters()
	{
		printers.clear()
	}

	private fun print(value: List<String>): String
	{
		val lastPrinter = printers.last()
		return lastPrinter(value)
	}
}

File: FizzBuzz/JavaScript/src/FizzBuzz.js
const filterJob = require('./FilterJob').filterJob
const printerJob = require('./PrinterJob').printerJob

class FizzBuzz {

    stringFor(input) {
	const result = this.process(input)
	return this.print(input, result)
    }

    process(input) {
	return filterJob.process(input)
    }

    print(input, result) {
	return printerJob.process(input, result)
    }
}

module.exports = { fizzBuzz: new FizzBuzz() }

